crucible-llvm/src/Lang/Crucible/LLVM/SimpleLoopInvariant.hs

```diff
@@ -0,0 +1,1151 @@
+import qualified Lang.Crucible.Simulator.RegMap as C
+import qualified Lang.Crucible.Simulator as C
+
+import qualified Lang.Crucible.LLVM.Bytes as C
+import qualified Lang.Crucible.LLVM.DataLayout as C
+import qualified Lang.Crucible.LLVM.MemModel as C
+import qualified Lang.Crucible.LLVM.MemModel.MemLog as C hiding (Mem)
+import qualified Lang.Crucible.LLVM.MemModel.Pointer as C
+import qualified Lang.Crucible.LLVM.MemModel.Type as C
+
+
+-- | A datatype describing the reason we are building an instance
+--   of the loop invariant.
+data InvariantPhase
+  = InitialInvariant
+  | HypotheticalInvariant
+  | InductiveInvariant
+
+-- | When live loop-carried dependencies are discovered as we traverse
+--   a loop body, new "widening" variables are introduced to stand in
+--   for those locations.  When we introduce such a varible, we
+--   capture what value the variable had when we entered the loop (the
+--   \"header\" value); this is essentially the initial value of the
+--   variable.  We also compute what value the variable should take on
+--   its next iteration assuming the loop doesn't exit and executes
+--   along its backedge.  This \"body\" value will be computed in
+--   terms of the the set of all discovered live variables so far.
+--   We know we have reached fixpoint when we don't need to introduce
+--   and more fresh widening variables, and the body values for each
+--   variable are stable across iterations.
+data InvariantEntry sym tp =
+  InvariantEntry
+  { headerValue :: W4.SymExpr sym tp
+  , bodyValue :: W4.SymExpr sym tp
+  }
+
+instance OrdF (W4.SymExpr sym) => OrdF (InvariantEntry sym) where
+  compareF x y = case compareF (headerValue x) (headerValue y) of
+    LTF -> LTF
+    EQF -> compareF (bodyValue x) (bodyValue y)
+    GTF -> GTF
+
+instance OrdF (InvariantEntry sym) => W4.TestEquality (InvariantEntry sym) where
+  testEquality x y = case compareF x y of
+    EQF -> Just Refl
+    _ -> Nothing
+
+-- | This datatype captures the state machine that progresses as we
+--   attempt to compute a loop invariant for a simple structured loop.
+data FixpointState p sym ext rtp blocks
+    -- | We have not yet encoundered the loop head
+  = BeforeFixpoint
+
+    -- | We have encountered the loop head at least once, and are in the process
+    --   of converging to an inductive representation of the live variables
+    --   in the loop.
+  | ComputeFixpoint (FixpointRecord p sym ext rtp blocks)
+
+    -- | We have found an inductively-strong representation of the live variables
+    --   of the loop. We are now executing the from the loop head one final time
+    --   to produce the proof obligations arising from the body of the loop,
+    --   the main inductive loop invariant obligation, and any obligations arising
+    --   from code following the loop exit.
+    | CheckFixpoint (FixpointRecord p sym ext rtp blocks)
+
+
+type ArrayTp = W4.BaseArrayType (C.EmptyCtx C.::> W4.BaseBVType 64) (W4.BaseBVType 8)
```

Comment:
This is a curious definition. What is its purpose?

Reply:
I was using it in some time signatures somewhere, but perhaps all the uses got deleted... I'll check.